### PR TITLE
feat(perf): Added new CLI flags for iterations + updated docs

### PR DESCRIPTION
**Pull request by maschad**
While integrating [the js impl of perf](https://github.com/libp2p/js-libp2p/pull/1604) I encountered some obstacles in running this infrastructure as well as adding and testing the _perf_ implementation. These are some improvements I thought would help new users of _perf_

Related https://github.com/libp2p/test-plans/issues/222

### CODE REVIEW

File: perf/README.md [hunk 1]
```diff
@@ -15,24 +15,49 @@ Benchmark results can be visualized with https://observablehq.com/@libp2p-worksp
...
-1. Save your public SSH key as the file `./terraform/modules/short_lived/files/perf.pub`; or generate a new key pair with `make ssh-keygen` and add it to your SSH agent with `make ssh-add`.
-2. `cd terraform/configs/local`
-3. `terraform init`
-4. `terraform apply`
-5. `CLIENT_IP=$(terraform output -raw client_ip)`
-6. `SERVER_IP=$(terraform output -raw server_ip)`
+1. Ensure you ssh agent is running, you can check that via
+```bash
+   eval "$(ssh-agent -s)"
+```
```
Comment (marten-seemann) [inline perf/README.md:31]: I don't think SSH requires any background tasks.

**Review by p-shahi — COMMENTED** [review]
drive by review

File: perf/README.md (hunk 1 above)
Comment (maschad) [reply perf/README.md:31]: If I try to run [make ssh-add](https://github.com/libp2p/test-plans/blob/master/perf/Makefile#L4) without running an ssh-agent I will get the error

```
Could not add identity "./terraform/modules/short_lived/files/perf": agent refused operation
make: *** [ssh-add] Error 1
```

Comment (marten-seemann) [reply perf/README.md:31]: It’s not a requirement for using SSH.
I don’t think we need to point out utterly trivial things.

File: perf/README.md [hunk 2]
```diff
@@ -26,13 +33,28 @@ Benchmark results can be visualized with https://observablehq.com/@libp2p-worksp
...
+
+### Build and run libp2p implementations
 
-_WARNING_: Running the perf tests might take a while.
+Given you have provisioned your infrastructure, you can now build and run the libp2p implementations on the AWS instances.
 
 1. `cd runner`
 2. `npm ci`
-3. `npm run start -- --client-public-ip $CLIENT_IP --server-public-ip $SERVER_IP`
+3.  `npm run start -- --client-public-ip $CLIENT_IP --server-public-ip $SERVER_IP --iterations 50`
```
[the remark this answers is not in the document]
Comment (maschad) [reply perf/README.md:56]: The intention was to highlight the usage of the `iterations` flag, but given that's [now addressed in a note](https://github.com/libp2p/test-plans/pull/258/files#diff-f67fd5d0f863534b4c355fab5e18cb17656ef46a4bb892f1fd885a36ce01e3e5R57) this probably isn't that useful anymore.

**Review by p-shahi — APPROVED** [review]

File: perf/runner/src/index.ts [hunk 3]
```diff
@@ -37,7 +37,7 @@ async function main(clientPublicIP: string, serverPublicIP: string) {
                  uploadBytes: 1,
                  downloadBytes: 1,
                  unit: "s",
-                 iterations: 100,
+                 iterations: iterations * 10,
```
Comment (mxinden) [inline perf/runner/src/index.ts:40]: I don't think this is intuitive. When I set `--iterations 10` I don't expect `100` iterations to run instead.

File: perf/README.md [hunk 4]
```diff
@@ -15,7 +15,14 @@ Benchmark results can be visualized with https://observablehq.com/@libp2p-worksp
 3. Wait for action run to finish and to push a commit to your branch.
 4. Visualize results on https://observablehq.com/@libp2p-workspace/performance-dashboard.
 
-## Running via local machine
+## Running with Terraform on AWS manually
+
+### Prerequisites
+
+- Terraform 1.5.5 or later
```
Comment (mxinden) [inline perf/README.md:22]: Note that I am using `terraform` `v1.5.4` without issues thus far.

File: perf/README.md [hunk 5]
```diff
@@ -26,13 +33,28 @@ Benchmark results can be visualized with https://observablehq.com/@libp2p-worksp
 5. `CLIENT_IP=$(terraform output -raw client_ip)`
 6. `SERVER_IP=$(terraform output -raw server_ip)`
 
-### Build and run implementations
+**Notes**
+- You may need to reset the infrastructure if you encounter any errors, you can do that by running `terraform destroy` and then `terraform apply`.
```
Comment (mxinden) [inline perf/README.md:37]: I don't think this is a good general hint. In the most cases terraform should be able to reconciliate any changes without a previous `terraform destroy`. On a high level, I am hesitant to give general terraform advice here. I think that is much better placed in the official terraform documentation.

File: perf/README.md [hunk 6]
```diff
@@ -41,27 +63,28 @@ _WARNING_: Running the perf tests might take a while.
...
-            - `--transport` (see `runner/versions.ts` for possible variants)
+            - `--server-address`
+            - `--transport` (see [`runner/versions.ts`](./runner/src/versions.ts#L7-L43) for possible variants)
             - `--upload-bytes` number of bytes to upload per stream.
             - `--download-bytes` number of bytes to download per stream.
           - Output
-            - Logging MUST go to stderr.
-            - Measurement output is printed to stdout as JSON in the form of:
+            - Logging MUST go to `stderr`.
+            - Measurement output is printed to **stdout** as JSON in the form of:
```
Comment (mxinden) [inline perf/README.md:83]: Nit pick. Either use monospace or bold.

File: perf/README.md [hunk 7]
```diff
@@ -26,13 +33,28 @@ Benchmark results can be visualized with https://observablehq.com/@libp2p-worksp
...
+**Notes**
+- You may need to reset the infrastructure if you encounter any errors, you can do that by running `terraform destroy` and then `terraform apply`.
+- While running terraform you may encounter the following error:
+  ```bash
+    Error: collecting instance settings: reading EC2 Launch Template versions: couldn't find resource
+    │
+    │   with module.short_lived_server[0].aws_instance.perf,
+    │   on ../../modules/short_lived/main.tf line 15, in resource "aws_instance" "perf":
+    │   15: resource "aws_instance" "perf" {
+  ```
```
Comment (mxinden) [inline perf/README.md:45]: Let's add some context here along the lines of the below.

> This implies that you haven't deployed the long-living infrastructure (e.g. launch template) on your AWS account. To do so along with each short-lived deployment, you can set the `long_lived_enabled` ...

**Review by mxinden — COMMENTED** [review]
I don't think this pull request was ready for merge. This is partially on me, due to my delayed review.

I don't think the `--iterations` flag is intuitive. See comment below. Also it is likely going away again with https://github.com/libp2p/test-plans/issues/261, as upload and download bandwidth tests will likely only run as a single iteration.

**Comment by mxinden** [comment]
How about replacing the `--iterations` flag with a `--testing` flag, which would reduce all benchmarks (ping, iperf, ...) to run a single time only. That would remove the counterintuitive 100 iterations on `--iterations 10` and is future proof for https://github.com/libp2p/test-plans/issues/261.

**Pull request merged**